Parser/Parser.cs

```diff
@@ -77,16 +117,76 @@ NBlock Block ()
       Expect (BEGIN);
       List<NStmt> stmts = new ();
       while (!Match (END)) { stmts.Add (Stmt ()); Match (SEMI); }
+      Match (SEMI);
       return new (stmts.ToArray ());
    }
 
    // write-stmt =  ( "writeln" | "write" ) arglist .
    NWriteStmt WriteStmt () 
       => new (Prev.Kind == WRITELN, ArgList ());
 
+   // read-stmt = "read" "(" identlist ")" .
+   NReadStmt ReadStmt () {
+      Expect (OPEN); Expect (IDENT); Expect (CLOSE);
+      return new (Prev);
+   }
+
    // assign-stmt = IDENT ":=" expr .
    NAssignStmt AssignStmt () 
       => new (PrevPrev, Expression ());
+
+   // call-stmt =  IDENT arglist .
+   NCallStmt CallStmt ()
+      => new (Prev, ArgList ());
+
+   // if-stmt =  "if" expression "then" statement [ "else" statement ] .
+   NIfStmt IfStmt () {
+      var ifExpr = Expression ();
+      Expect (THEN);
+      List<NStmt> thenStmts = new (), elseStmts = new ();
+      if (Match (BEGIN))
+         do { thenStmts.Add (Stmt ()); Match (SEMI); } while (!Match (END));
+      else thenStmts.Add (Stmt ());
+      Match (SEMI);
+      if (Match (ELSE))
+         if (Match (BEGIN))
+            do { elseStmts.Add (Stmt ()); Match (SEMI); } while (!Match (END));
+         else elseStmts.Add (Stmt ());
+      return new NIfStmt (ifExpr, new (thenStmts.ToArray ()), new (thenStmts.ToArray ()));
+   }
+
+   // while-stmt =  "while" expression "do" statement .
+   NWhileStmt WhileStmt () {
+      var expr = Expression ();
+      Expect (DO);
+      List<NStmt> stmts = new ();
+      if (Match (BEGIN))
+         do { stmts.Add (Stmt ()); Match (SEMI); } while (!Match (END));
+      else stmts.Add (Stmt ());
+      return new (expr, new (stmts.ToArray ()));
+   }
+
+   // repeat-stmt =  "repeat" statement { ";" statement} "until" expression. 
+   NRepeatStmt RepeatStmt () {
+      List<NStmt> stmts = new ();
+      while (!Match (UNTIL)) { stmts.Add (Stmt ()); Match (SEMI); }
+      var expr = Expression ();
+      Expect (SEMI);
+      return new (expr, stmts.ToArray ());
+   }
+
+   // for-stmt =  "for" IDENT ":=" expression ( "to" | "downto" ) expression "do" statement.
+   NForStmt ForStmt () {
+      var ident = Expect (IDENT);
+      Expect (ASSIGN); var from = Expression ();
+      Expect (TO, DOWNTO); var to = Expression ();
+      Expect (DO);
+      List<NStmt> stmts = new ();
+      if (Match (BEGIN))
+         do { stmts.Add (Stmt ()); Match (SEMI); } while (!Match (END));
+      else stmts.Add (Stmt ());
+      return new (ident, from, to, stmts.ToArray ());

```

Comment:
According to the language declaration, it takes only one statement.
So, it can just be:
```
NForStmt ForStmt () {
      var ident = Expect (IDENT); Expect (ASSIGN); var exp = Expression ();
      var ident2 = Expect (TO, DOWNTO); var exp2 = Expression ();
      Expect (DO);
      return new (ident, exp, ident2.Kind == TO, exp2, Stmt ());
   }
```